### PR TITLE
Unified Attention Sparse MLA FP8

### PR DESCRIPTION
**Pull request by anhminhnguyenhoang**
Co-authors: @anhminhnguyenhoang

## Motivation

Extends the existing Triton **`unified_attention_sparse_mla`** kernel (dense top-k, BF16 only) with:

- **CSR input form** `(kv_indptr, kv_indices)` alongside the existing dense `[seq_len, TOP_K]` matrix — variable per-query length, no padding, matches the selector output shape used by SGLang's `nsa_backend`.
- **3D split-K path** for low-batch decode where the existing 2D grid leaves most CUs idle — gives a **1.4–2.9× decode-latency win** over `aiter.mla.mla_decode_fwd` at a generic DSA decode shape `heads=16, lora=512, rope=64, block=64, top_k=2048`.
- **FP8 (e4m3) KV-cache support folded into the main wrapper** — per-tensor `q/k/v_scale` flow through the 2D and 3D paths without a separate `_fp8` shim.

## Technical Details

`unified_attention_sparse_mla()` dispatches to one of two kernel families chosen at call time:

1. **2D path (`_kernel_unified_attention_sparse_mla_2d`)** -- the pre-existing dense top-k kernel body, extended to also serve the new CSR input form via `USE_CSR: tl.constexpr`. The only divergent region between the two forms is the per-tile KV-index fetch, so Triton specialization yields two AMDGCN variants with no runtime branch cost. Used when `total_num_q_blocks >= _NUM_CU_HINT` (grid already fills the GPU). Vestigial `block_tables_ptr` / `block_table_stride` params dropped -- both 2D bodies derive `(page, slot)` directly from `pos // BLOCK_SIZE` / `pos % BLOCK_SIZE`.
2. **3D split-K path (`_kernel_unified_attention_sparse_mla_csr_3d` + `_kernel_unified_attention_sparse_mla_csr_reduce`)** -- **new** in this PR. For low-batch decode where the existing 2D grid would leave most CUs idle. Grid is `(num_q_blocks, NUM_SEGMENTS_PER_SEQ)` so a single-batch decode fills the GPU instead of running 1 CTA on 304 CUs. `NUM_SEGMENTS_PER_SEQ` is chosen from a CU-count hint and rounded to a power of two for the LSE reduce kernel layout.

Shared softmax / dtype work added in this PR (applies to both paths):

- Softmax via `tl.math.exp2` with `RCP_LN2` folded into `qk_scale` once (replaces `tl.math.exp`).
- Fused `tl.dot(Q_lora, K_lora, acc=S)` over the rope dot.
- `PRELOAD_V=True` to overlap V load with softmax.
- FP8 KV-cache plumbing: per-tensor `q_scale / k_scale / v_scale` (scalar tensor or python float; coerced to 0-d float32) thread through 2D dense, 2D CSR, and 3D split-K. `K_SCALE` folds into `qk_scale`; `V_SCALE` applies post-softmax in the epilogue, so the 3D reduce kernel stays scale-agnostic. Per-tile `.to(Q.dtype)` casts promote FP8 K_rope / V_lora to BF16 before `tl.dot`.

**Dispatch policy** (`aiter/ops/triton/attention/unified_attention_sparse_mla.py`):

- 3D split-K when CSR inputs are passed, `total_num_q_blocks < UNIFIED_ATTENTION_SPARSE_MLA_NUM_CU`, and `max_sparse_len >= UNIFIED_ATTENTION_SPARSE_MLA_SPLIT_K_THRESHOLD` (default 1024). FP8 uses the same policy as BF16.
- 2D otherwise. Dense top-k path is preserved bit-compatible.

`max_sparse_len` is an optional argument -- pass it explicitly to skip a device->host `.item()` per decode step.

**Tuned defaults** (autotune sweep on `heads=16, lora=512, rope=64, block=64, top_k=2048` decode shapes, MI355X -
- generic DSA-style decode):

- 3D path: `TILE_SIZE=32, num_warps=8, num_stages=2, PRELOAD_V=True, waves_per_eu=2`.
- 2D CSR path: `TILE_SIZE=64, num_warps=4, num_stages=1, PRELOAD_V=True, waves_per_eu=2` - only the CSR launch site uses these.

**Layout**:

- Kernels: `aiter/ops/triton/_triton_kernels/attention/unified_attention_sparse_mla.py` -- `_kernel_unified_attention_sparse_mla_2d` (with `USE_CSR` constexpr branch + FP8 scale args), new `_kernel_unified_attention_sparse_mla_csr_3d` and `_kernel_unified_attention_sparse_mla_csr_reduce`, plus three autotuners: `_2d_topk_autotuner`, `_2d_csr_autotuner`, `_3d_csr_autotuner`. The two 2D autotuners wrap the same merged kernel with distinct cache keys so per-path tuning state stays isolated.
- API: `aiter/ops/triton/attention/unified_attention_sparse_mla.py` -- `unified_attention_sparse_mla()` wrapper extended for CSR input, 3D dispatch, FP8 scale coercion. `BLOCK_M=16` is unchanged from prior work.
- Tests: `op_tests/triton_tests/attention/test_unified_attention_sparse_mla.py` -- extended to cover the new CSR + 3D + FP8 paths.
- Benchmark: `op_tests/op_benchmarks/triton/bench_unified_attention_sparse_mla.py` -- vs ASM `mla_decode_fwd`; supports `--dtype {bf16,fp8}`, `--sparse-pattern {nsa,dsa}` (DSA shuffles each row's indices to simulate the lightning-indexer's scattered selection), `--validate` (torch reference), `--skip-mla-decode` (when aiter HIP isn't built).

**Env knobs** (read by the wrapper):

| Env var | Default | Effect |
|---|---|---|
| `UNIFIED_ATTENTION_SPARSE_MLA_SPLIT_K_THRESHOLD` | `1024` | Minimum `max_sparse_len` to enable 3D split-K. Lower for shorter-K decode shapes that still underfill the GPU. |
| `UNIFIED_ATTENTION_SPARSE_MLA_DISABLE_SPLIT_K` | `0` | Force-disable the 3D path for A/B comparisons. |
| `UNIFIED_ATTENTION_SPARSE_MLA_NUM_CU` | `256` | CU-count hint used to size `NUM_SEGMENTS_PER_SEQ`. MI355X has 304 CUs; default leaves headroom for waves. |
| `UNIFIED_ATTENTION_SPARSE_MLA_AUTOTUNE` | unset | When set, runs the three Triton autotuners instead of baked-in defaults -- slow first call, useful for re-sweeping on a new shape. |

## Test Plan

`op_tests/triton_tests/attention/test_unified_attention_sparse_mla.py` covers both the 2D and 3D dispatch paths under both input forms (dense top-k and CSR) at BF16 and FP8 KV-cache, against a torch reference. The `--validate` flag in the bench script provides an additional online correctness check. After the 2D merge, **178/178 test cases pass**; CSR microbench parity within +/-3% at every shape.

## Test Setup

image `rocm/sgl-dev:v0.5.10.post1-rocm720-mi35x-20260427` (pre-installed `aiter` lives at `/sgl-workspace/aiter` ). 
Replace `your_NTID` as needed:

```
docker run -it --name sparse-mla-dev --device=/dev/kfd --device=/dev/dri \
  --group-add video --ipc=host --cap-add=SYS_PTRACE --security-opt seccomp=unconfined \
  -v /home/your_NTID/aiter_ua:/workspace/aiter_ua \
  -w /sgl-workspace/aiter \
  rocm/sgl-dev:v0.5.10.post1-rocm720-mi35x-20260427 \
  bash

pip uninstall amd-aiter
cd /workspace/aiter_ua && pip install -e .
pytest op_tests/triton_tests/attention/test_unified_attention_sparse_mla.py -v -s

# microbench (BF16)
python op_tests/op_benchmarks/triton/bench_unified_attention_sparse_mla.py \
    --batch 1 --sq 1 --sk 2048 \
    --heads 16 --lora-dim 512 --rope-dim 64 --block-size 64 \
    --top-k 2048 --warmup 25 --rep 100 --validate

# microbench (FP8 KV-cache)
python op_tests/op_benchmarks/triton/bench_unified_attention_sparse_mla.py \
    --batch 1 --sq 1 --sk 2048 \
    --heads 16 --lora-dim 512 --rope-dim 64 --block-size 64 \
    --top-k 2048 --warmup 25 --rep 100 --validate --dtype fp8
```

Sweep `--batch` in `{1, 8, 32, 64}` to see how the 3D split-K advantage tapers as the 2D grid naturally fills the GPU. Add `--sparse-pattern dsa` to A/B the kernel's gather throughput against block-clustered (NSA) vs uniformly scattered (DSA) index distributions on the same row sets.

### gfx950 (MI355X) -- GLM-5 decode shape sweep

Environment: Ubuntu 22.04.5, ROCm 7.2.0, PyTorch 2.9.1+rocm7.2.0, **Triton 3.6.0**, image `rocm/sgl-dev:v0.5.10.post1-rocm720-mi35x-20260427`. Shape: `heads=16, lora=512, rope=64, block=64, sk=2048, top_k=2048`, BF16 KV-cache. CSR is the Triton path under test; baseline is `aiter.mla.mla_decode_fwd` (ASM).

| batch | csr_ms | mla_decode_fwd_ms | csr speedup vs mla |
|---|---|---|---|
| 1  | 0.0365 | 0.1074 | **2.94x** |
| 8  | 0.0311 | 0.0997 | **3.21x** |
| 32 | 0.0424 | 0.0982 | **2.32x** |
| 64 | 0.0729 | 0.1040 | **1.43x** |

The 3D split-K restructure is the dominant lever -- at batch=1 it accounts for an ~8x local speedup over a 2D-only version.
Speedup compresses at batch=64 (CSR scales ~linearly with work, `mla_decode_fwd` stays flat -- they cross around batch >= 128 at this sk).

### NSA vs DSA sparsity-pattern A/B (heads=128, MI355X)

`--sparse-pattern {nsa,dsa}` toggles between block-clustered indices (test harness default) and uniformly scattered indices that simulate DeepSeek V3.2-Exp's lightning-indexer token-level selection. **Within noise** at every batch -- the kernel's int32 index-load path isn't coalesced enough for clustering to matter:

| batch | nsa csr_ms | dsa csr_ms | mla_decode_fwd_ms |
|---|---|---|---|
| 1  | 0.0506 | 0.0391 | 0.124 |
| 8  | 0.0740 | 0.0739 | 0.109 |
| 32 | 0.3064 | 0.3058 | 0.103 |
| 64 | 0.4418 | 0.4382 | 0.106 |

At `heads=128`, batch <= 8 the 3D CSR path wins (~2.4x over `mla_decode_fwd` at batch=1). Batch >= 32 falls into the 2D CSR path and currently loses ~3-4x to ASM -- the kernel's per-CTA work at heads=128 high-batch outruns occupancy.

## Known Limitations

- **`heads=128 batch >= 32`** in the 2D CSR path is ~3-4x behind ASM `mla_decode_fwd` even after the re-tune. Closing the gap likely requires a kernel restructure (better int32 index-load coalescing or `BLOCK_M` re-tune), not config tuning.
- **`BLOCK_M=16`** is hardcoded in the wrapper; changing it requires re-tuning the autotuner configs.
- **FP8 at higher batch (>=32)** is marginally slower than BF16 (~5%) -- the FP8 -> BF16 promotion casts cost more than the K-cache bandwidth savings buy back once the kernel is no longer bandwidth-bound. FP8 still wins at batch=1 where the 3D path matters most.

### CODE REVIEW

**Comment by zufayu** [comment]
@anhminhnguyenhoang This PR has been inactive for over 2 months. Is the Unified Attention Sparse MLA FP8 still being pursued? If you still plan to pursue it, please rebase onto current main and push an update. If there's no response within ~2 weeks, we'll close this to keep the PR queue clean — it can always be reopened or resubmitted.
